turbopack/crates/turbo-tasks/src/graph/adjacency_map.rs

```diff
@@ -174,6 +189,43 @@ where
     }
 }
 
+pub struct IntoBreadthFirstEdges<T>
+where
+    T: Eq + std::hash::Hash + Clone,
+{
+    adjacency_map: HashMap<T, Vec<T>>,
+    stack: VecDeque<(Option<T>, T)>,
+    visited: HashSet<T>,
+}
+
+impl<T> Iterator for IntoBreadthFirstEdges<T>
+where
+    T: Eq + std::hash::Hash + Clone,
+{
+    type Item = (Option<T>, T);
+
+    fn next(&mut self) -> Option<Self::Item> {
+        let (parent, current) = self.stack.pop_front()?;
+
+        let Some(neighbors) = self.adjacency_map.get(&current) else {
+            self.visited.insert(current.clone());
+            return Some((parent, current));
+        };
+
+        if !self.visited.contains(&current) {
+            self.stack.extend(
+                neighbors
+                    .iter()
+                    .rev()
+                    .map(|neighbor| (Some(current.clone()), neighbor.clone())),
+            );
+        }
+        self.visited.insert(current.clone());
```

Comment:
Insert returns true only on successful insert

```suggestion
        if self.visited.insert(current.clone()) {
            self.stack.extend(
                neighbors
                    .iter()
                    .rev()
                    .map(|neighbor| (Some(current.clone()), neighbor.clone())),
            );
        }
```